### PR TITLE
[mlir][ArmSME] Add optional padding and mask operands to tile_load

### PR DESCRIPTION
**Pull request by c-rhodes**
Padding and mask are optional, but if one is specified both must be specified. This is consistent with vector.transfer_read.

### CODE REVIEW

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td [hunk 1]
```diff
@@ -231,7 +231,24 @@ def ZeroOp : ArmSME_Op<"zero", [Pure]> {
...
+    "::llvm::cast<VectorType>($_self).getElementType()",
+    "!getPadding() || std::equal_to<>()"
+  >,
+  TypesMatchWith<
+    "mask has i1 element type and same shape as result (if present)",
+    "result", "mask",
+    "VectorType("
+      "VectorType::Builder("
+        "::llvm::cast<mlir::VectorType>($_self)"
+      ").setElementType(IntegerType::get($_self.getContext(), 1)))",
```
Comment (dcaballe) [inline mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:248]: any chances you can check this without building a new type?

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td [hunk 2]
```diff
@@ -242,6 +259,16 @@ def TileLoadOp : ArmSME_Op<"tile_load"> {
...
 
+    An optional SSA value `padding` of the same elemental type as the MemRef is
+    provided to specify a fallback value in the case of masking.
+
+    An optional SSA value `mask` may be specified to mask out elements read
+    from the MemRef. The `mask` type is an `i1` vector with a shape that
+    matches how elements are read from the MemRef. Elements whose corresponding
+    mask element is `0` are masked out and replaced with `padding`.
+
+    If either `padding` or `mask` are specified, both must be specified.
```
Comment (dcaballe) [inline mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:270]: is this being verified + tested?

**Review by dcaballe — APPROVED** [review]

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td (hunk 1 above)
Comment (MacDue) [reply mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:248]: This is not just checking the type, it's used to infer the type of the mask from the result. That requires constructing the type (I believe).

File: mlir/test/Dialect/ArmSME/invalid.mlir [hunk 3]
```diff
@@ -90,10 +106,38 @@ func.func @arm_sme_move_vector_to_tile_slice_f32__bad_vector_type(%vector : vect
...
+
+//===----------------------------------------------------------------------===//
+// arm_sme.tile_load
+//===----------------------------------------------------------------------===//
+
+// -----
+
+func.func @arm_sme_tile_load__bad_padding_type(%src : memref<?x?xf64>, %pad : f32, %mask : vector<[2]x[2]xi1>) {
+  %c0 = arith.constant 0 : index
+  // expected-note@-2 {{prior use here}}
```
Comment (banach-space) [inline mlir/test/Dialect/ArmSME/invalid.mlir:129]: [nit] We can probably skip this line

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td (hunk 2 above)
Comment (banach-space) [reply mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:270]: Yeah, i think that's missing. If this is to require a custom verifier then I'd be happy to be implemented in a separate patch (to unblock tings).

**Review by banach-space — APPROVED** [review]
LGTM, ta!

The extra test for the verifier (to make sure that neither `mask` nor `pad` can be used without the other) could be added separately.

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td (hunk 2 above)
Comment (MacDue) [reply mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:270]: This is enforced by the syntax (with only allows specifying both or neither). You could also verify this constraint with: 
```
    PredOpTrait<
      "both `lhsMask` and `rhsMask` should be provided or neither",
      CPred<"bool(getPadding()) == bool(getMask())">>,
```

**Comment by c-rhodes** [comment]
> LGTM, ta!
> 
> The extra test for the verifier (to make sure that neither `mask` nor `pad` can be used without the other) could be added separately.

Thanks for reviewing. I've added the test.

**Pull request merged**